libvips/foreign/dzsave.c

```diff
@@ -2532,87 +2207,80 @@ vips_foreign_save_dz_build( VipsObject *object )
-				"%s", dz->dirname );
-		else
-			vips_snprintf( name, VIPS_PATH_MAX, 
-				"%s/%s", dz->dirname, dz->basename ); 
-
-		if( !(out = (GsfOutput *) 
-			gsf_output_dir_new( name, NULL )) ) {
-			return( -1 );
-		}
-	
-		dz->tree = vips_gsf_tree_new( out, 0 );
-}
-		break;
-
-	case VIPS_FOREIGN_DZ_CONTAINER_ZIP:
-	case VIPS_FOREIGN_DZ_CONTAINER_SZI:
-{
-		GsfOutput *zip;
-		GsfOutput *out2;
-		GError *error = NULL;
+		VIPS_SETSTR( dz->dirname, "" );
 
 		/* We can have dzsave("x.zip", container="fs"), ie. zip output
 		 * from write to file. Make a target if we need one.
 		 */
 		if( !dz->target ) {
-			if( !(dz->target = 
+			if( !(dz->target =
 				vips_target_new_to_file( dz->filename )) )
 				return( -1 );
 		}
 
-		/* Can be memory, a file (not a directory tree), pipe, etc.
+		/* Allocate and initialize a new archive.
 		 */
-		dz->out = gsf_output_target_new( dz->target );
+		if( !(dz->archive = archive_write_new()) )
+			return( -1 );
 
-		if( !(zip = (GsfOutput *) 
-			gsf_outfile_zip_new( dz->out, &error )) ) {
-			vips_g_error( &error );
+		/* Set format to zip.
+		 */
+		if( archive_write_set_format( dz->archive,
+			ARCHIVE_FORMAT_ZIP ) != ARCHIVE_OK ) {
+			archive_write_free( dz->archive );
 			return( -1 );
 		}
 
-		/* Make the base directory inside the zip. All stuff goes into
-		 * this. 
+		/* Remap compression=-1 to compression=6.
 		 */
-		out2 = gsf_outfile_new_child_full( (GsfOutfile *) zip, 
-			dz->basename, TRUE,
-			"compression-level", GSF_ZIP_STORED, 
-			NULL );
-
-#ifndef HAVE_GSF_DEFLATE_LEVEL
-		if( dz->compression > 0 ) {
-			g_warning( "%s", 
-				_( "deflate-level not supported by libgsf, "
-				"using default compression" ) ); 
-			dz->compression = -1;
-		}
-#endif /*HAVE_GSF_DEFLATE_LEVEL*/
+		if( dz->compression == -1 )
+			dz->compression = 6; /* Z_DEFAULT_COMPRESSION */
 
-		dz->tree = vips_gsf_tree_new( out2, dz->compression );
+		/* Deflate compression requires libarchive >= v3.2.0.
+		 * https://github.com/libarchive/libarchive/pull/84
+		 */
+#if ARCHIVE_VERSION_NUMBER >= 3002000
+		/* Set deflate compression level.
+		 */
+		char compression_string[2] = { '0' + dz->compression, 0 };
+		if( archive_write_set_format_option( dz->archive, "zip",
+			"compression-level", compression_string ) != ARCHIVE_OK ) {
+			archive_write_free( dz->archive );
+			return( -1 );
+		}
+#else
+		if( dz->compression > 0 )
+			g_warning(
+				"%s: libarchive >= v3.2.0 required for Deflate compression",
+				class->nickname );
+#endif
 
-		/* Note the thing that will need closing up on exit.
+		/* Do not pad last block.
 		 */
-		dz->tree->container = zip; 
-}
-		break;
+		if( archive_write_set_bytes_in_last_block( dz->archive, 1 ) ) {
+			archive_write_free( dz->archive );
+			return( -1 );
+		}
 
-	default:
-		g_assert_not_reached();
+		/*if( archive_write_add_filter( dz->archive,
+			ARCHIVE_FILTER_NONE ) != ARCHIVE_OK ) {
+			archive_write_free( dz->archive );
+			return( -1 );
+		}*/
```

Comment:
Can this commented-out block be removed?

Reply:
aee3605c9cace5f5941cbdbf247113e3098c89fc